### PR TITLE
Tools 2092 (asbackup) uninitialized record check before move records

### PR DESCRIPTION
**Pull request by atefehmohseni**

### CODE REVIEW

File: src/utils.c [hunk 1]
```diff
@@ -1798,7 +1798,13 @@ bool
 as_key_move(as_key* dst, as_key* src)
 {
 	*dst = *src;
-	if (as_load_uint32(&src->value.integer._.count) > 1) {
+	if (!src->valuep) {
+		return true;
+	}
+	if (as_load_uint32(&src->valuep->integer._.count) > 1 || \
+		as_load_uint32(&src->valuep->string._.count) > 1 || \
+		as_load_uint32(&src->valuep->bytes._.count) > 1 ) {
```
Comment (dwelch-spike) [inline src/utils.c]: Per Brian's information lets change this back to a single check on integer._.count.

File: test/unit/test_utils.c [hunk 2]
```diff
@@ -367,6 +367,73 @@ START_TEST(test_timespec_diff_1ms_overflow)
 }
 END_TEST
 
+
+START_TEST(test_as_key_move_an_empty_rec)
```
Comment (dwelch-spike) [inline test/unit/test_utils.c]: nit: "empty_rec" should maybe be "empty_key"

File: test/unit/test_utils.c [hunk 3]
```diff
@@ -367,6 +367,73 @@ START_TEST(test_timespec_diff_1ms_overflow)
...
+{
+	as_key* test_key = (as_key *) cf_malloc(sizeof(as_key));
+	
+	ck_assert(as_key_move(test_key, test_key));
+}
+END_TEST
+
+START_TEST(test_as_key_move_a_not_empty_int_key)
```
Comment (dwelch-spike) [inline test/unit/test_utils.c]: nit: maybe "not_empty" could be "populated"

File: test/unit/test_utils.c [hunk 4]
```diff
@@ -367,6 +367,73 @@ START_TEST(test_timespec_diff_1ms_overflow)
 }
 END_TEST
 
+
+START_TEST(test_as_key_move_an_empty_rec)
+{
+	as_key* test_key = (as_key *) cf_malloc(sizeof(as_key));
```
Comment (dwelch-spike) [inline test/unit/test_utils.c]: I see Clayton had defined a `test_malloc` function that is used in other tests. I think we should use that here. I think the memory allocated by these should also be freed in the same tests. We have a common citrusleaf file called alloc.h that defines wrapped versions of these prefixed with `cf_`. I think using `cf_free` would be good. So `test_malloc()` and `cf_free()`.

File: src/utils.c [hunk 5]
```diff
@@ -1798,7 +1798,11 @@ bool
 as_key_move(as_key* dst, as_key* src)
 {
 	*dst = *src;
-	if (as_load_uint32(&src->value.integer._.count) > 1) {
+	if (!src->valuep) {
+		return true;
+	}
+	if (as_load_uint32(&src->valuep->integer._.count) > 1) {
```
Comment (dwelch-spike) [inline src/utils.c]: nit: we usually put an empty line between if statements. Unless there is something else specified in Clayton's style guide.

**Review by dwelch-spike — CHANGES_REQUESTED** [review]
I have a few comments and suggestions. Let me know what you think.

File: test/unit/test_utils.c (hunk 4 above)
Comment (atefehmohseni) [reply test/unit/test_utils.c]: thanks for the info. Could you please check out the order of calling "cf_free" in the latest commit?

**Review by dwelch-spike — APPROVED** [review]
LGTM but lets wait for actions tests to finish before merging.

**Pull request merged**